src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/InterceptedProjectProperties/ReferencesPage/ImportedNamespacesValueProvider.cs

```diff
@@ -0,0 +1,132 @@
+        });
+
+        return existingImports;
+    }
+    
+    private async Task<string> GetSelectedImportStringAsync()
+    {
+        StringBuilder sb = new StringBuilder();
+        string projectName = Path.GetFileNameWithoutExtension(_configuredProject.UnconfiguredProject.FullPath);
+        bool containsProjectName = false;
+
+        ImmutableArray<string> existingImports = await GetSelectedImportsAsync();
+        
+        foreach (string? import in existingImports)
+        {
+            if (!string.IsNullOrEmpty(import))
+            {
+                if (string.Equals(import, projectName))
+                {
+                    containsProjectName = true;
+                }
+                sb.Append($"{import};");
+            }
+        }
+
+        if (!containsProjectName)
+        {
+            sb.Append($"{projectName};");
+        }
+
+        sb.Remove(sb.Length - 1, 1);
+        return sb.ToString();
+    }
+    
+    public override Task<string> OnGetUnevaluatedPropertyValueAsync(string propertyName, string unevaluatedPropertyValue, IProjectProperties defaultProperties)
+    {
+        return GetSelectedImportStringAsync();
+    }
+
+    public override Task<string> OnGetEvaluatedPropertyValueAsync(string propertyName, string evaluatedPropertyValue, IProjectProperties defaultProperties)
+    {
+        return GetSelectedImportStringAsync();
+    }
+
+    public override async Task<string?> OnSetPropertyValueAsync(string propertyName, string unevaluatedPropertyValue, IProjectProperties defaultProperties, IReadOnlyDictionary<string, string>? dimensionalConditions = null)
+    {
+        await _threadingService.SwitchToUIThread();
+        
+        string[] newImportsToSet = unevaluatedPropertyValue.Split(';');
+        // delete existing imports that aren't in unevaluatedPropertyValue
+        
+        // add imports that are in unevaluatedPropertyValue but not in _imports
+        var importsToAdd = newImportsToSet.ToHashSet();
+        importsToAdd.Remove(Path.GetFileNameWithoutExtension(_configuredProject.UnconfiguredProject.FullPath));
+
+        foreach (string import in await GetSelectedImportsAsync())
+        {
+            if (!newImportsToSet.Contains(import))
+            {
+                try
+                {
+                    await _projectAccessor.OpenProjectForWriteAsync(_configuredProject, project =>
+                    {
+                        Microsoft.Build.Evaluation.ProjectItem importProjectItem = project.GetItems("Import")
+                            .First(i => string.Equals(import, i.EvaluatedInclude, StringComparisons.ItemNames));
+
+                        if (importProjectItem.IsImported)
+                        {
+                            _specialImports.Add(import);
+                        }
+                        
+                        project.RemoveItem(importProjectItem);
+                    });
+
+                }
+                catch (Exception)
```

Comment:
Catching all exceptions can mask serious problems. CPS has an extension for filtering out critical exceptions in cases like this. If we don't have access to it, we should probably port it and use it. Or change this to catch more specific exception type(s).

Reply:
Good point. Only catching ArgumentException/InvalidOperationException now